File: app/src/main/java/com/lifedawn/bestweather/findaddress/GoogleFindPlaceProcessing.java
package com.lifedawn.bestweather.findaddress;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.lifedawn.bestweather.data.remote.retrofit.client.RestfulApiQuery;
import com.lifedawn.bestweather.data.remote.retrofit.client.RetrofitClient;
import com.lifedawn.bestweather.data.remote.retrofit.parameters.google.GoogleFindPlaceParameter;
import com.lifedawn.bestweather.data.remote.retrofit.responses.google.placesearch.GooglePlaceSearchResponse;
import com.lifedawn.bestweather.data.remote.retrofit.callback.JsonDownloader;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class GoogleFindPlaceProcessing {
	public static GooglePlaceSearchResponse getGooglePlaceSearchResponse(String response) {
		return new Gson().fromJson(response, GooglePlaceSearchResponse.class);
	}

	public static void findPlace(String input, JsonDownloader callback) {
		RestfulApiQuery restfulApiQuery = RetrofitClient.getApiService(RetrofitClient.ServiceType.GOOGLE_PLACE_SEARCH);
		GoogleFindPlaceParameter parameter = new GoogleFindPlaceParameter(input);
		Call<JsonElement> call = restfulApiQuery.getFindPlaceSearch(parameter.getMap());

		call.enqueue(new Callback<JsonElement>() {
			@Override
			public void onResponse(Call<JsonElement> call, Response<JsonElement> response) {
				if (response.body() != null) {
					GooglePlaceSearchResponse googlePlaceSearchResponse =
							getGooglePlaceSearchResponse(response.body().toString());
					if (googlePlaceSearchResponse.getStatus().equals("OK")) {
						callback.onResponseResult(response, googlePlaceSearchResponse, response.body().toString());
					} else {
						callback.onResponseResult(new Exception());
					}
				} else {
					callback.onResponseResult(new Exception());
				}
			}

			@Override
			public void onFailure(Call<JsonElement> call, Throwable t) {
				callback.onResponseResult(t);
			}
		});
	}
}

File: app/src/main/java/com/lifedawn/bestweather/ui/findaddress/GoogleFindPlaceProcessing.kt
package com.lifedawn.bestweather.ui.findaddress

import com.lifedawn.bestweather.data.remote.retrofit.parameters.google.GoogleFindPlaceParameterRest
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

object GoogleFindPlaceProcessing {
    fun getGooglePlaceSearchResponse(response: String?): GooglePlaceSearchResponse {
        return Gson().fromJson(response, GooglePlaceSearchResponse::class.java)
    }

    fun findPlace(input: String?, callback: JsonDownloader) {
        val restfulApiQuery: RestfulApiQuery = RetrofitClient.getApiService(RetrofitClient.ServiceType.GOOGLE_PLACE_SEARCH)
        val parameter: GoogleFindPlaceParameterRest = GoogleFindPlaceParameterRest(input)
        val call: Call<JsonElement> = restfulApiQuery.getFindPlaceSearch(parameter.getMap())
        call.enqueue(object : Callback<JsonElement?> {
            public override fun onResponse(call: Call<JsonElement?>, response: Response<JsonElement?>) {
                if (response.body() != null) {
                    val googlePlaceSearchResponse: GooglePlaceSearchResponse = getGooglePlaceSearchResponse(response.body().toString())
                    if (googlePlaceSearchResponse.getStatus().equals("OK")) {
                        callback.onResponseResult(response, googlePlaceSearchResponse, response.body().toString())
                    } else {
                        callback.onResponseResult(Exception())
                    }
                } else {
                    callback.onResponseResult(Exception())
                }
            }

            public override fun onFailure(call: Call<JsonElement?>, t: Throwable) {
                callback.onResponseResult(t)
            }
        })
    }
}